feat(cli): add command to list installed browsers

niba opened the pull request:
Add a new CLI command to list all browser installations managed by Playwright. Fixes #34183

This command shows browsers installed by the _current_ Playwright instance
`npx playwright list`

![image](https://github.com/user-attachments/assets/97a9091c-1e5c-462e-8d5f-2ed7c1a7cadb)

To list browsers installed by _all_ Playwright instances, use the `--all` flag:

`npx playwright list --all`

![image](https://github.com/user-attachments/assets/5903ff34-5906-47dd-816f-6d4e5bfd1e94)

niba committed:
feat(cli): add command to list installed browsers (#34183)

niba commented:
@microsoft-github-policy-service agree

dgozman commented on packages/playwright-core/src/cli/program.ts:
```diff
@@ -199,6 +199,29 @@ Examples:
     Install custom browsers, supports ${suggestedBrowsersToInstall()}.`);
 
+program
+    .command('list')
```
I see that issues talks about `npx playwright install --list` instead of `npx playwright list`. Generally, we'd prefer everything about browsers to be under `install/uninstall`.

niba replied:
Sorry, that's my fault. I should have shared my thought process in the PR description. 
Initially, I started implementing the ⁠`--list` option. Then, I discovered that it is possible to list installations for all instances of Playwright, and I realized the ⁠`--all` flag could be useful.  I felt that adding ⁠`--all `and ⁠`--list `to the install command would be excessive, especially since ⁠`--all `only works in combination with ⁠`--list`.
That's why I decided to implement standalone list
This approach also follows the same pattern as the remove command:
	•	⁠`npx playwright remove (--all)`
	•	`⁠npx playwright list (--all)`

If you still prefer --list and --all flags I can change it.

dgozman commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1257,15 +1317,8 @@ export class Registry {
-              // All new applications have a marker file right away.
-              (browserName !== 'firefox' && browserName !== 'chromium' && browserName !== 'webkit');
-          if (!shouldHaveMarkerFile || (await existsAsync(browserDirectoryToMarkerFilePath(usedBrowserPath))))
+
```
Instead of copying pretty much half of this method, perhaps we can extract the "step 1" from this function into a separate one and reuse it?

niba replied:
I'm a bit lost. I extracted a part of the code that verifies the marker file into a standalone function ⁠`_validateMarkerFile`. I did this to reuse the same code in the remove and list functions. 
What do you mean by saying 'step 1'?

niba replied:
I don't know what to do with this one. I think I did it exactly as you described in my first commit. Can you take a look again?

niba replied:
I extracted as much as possible, thanks

dgozman commented on packages/playwright-core/src/cli/program.ts:
```diff
@@ -199,6 +199,29 @@ Examples:
+    .action(async (options: { all?: boolean }) => {
+      await registry.list(!!options.all).then(browsersInfo => {
+        for (const info of browsersInfo) {
+          const whichInstanceLog = ` (${info.currentInstance ? 'CURRENT' : 'OTHER'})`;
```
I kinda find `CURRENT/OTHER` being very much in my face. Can we tone it down somehow? Perhaps not show "other" at all, because we print the target directory anyway, and only mark current as `(current)` when `--all` is not passed?

dgozman reviewed (commented):
Thank you for the PR! I left some higher-level comments for now.

dgozman commented:
@niba Do you plan to finish this one? We've discussed this again with the team, and think that this should be a single command `npx playwright install --list` that will list browsers from all playwright installations, and highlight the current one similar to how you do it.

niba commented:
@dgozman  I've been waiting for clarification about the CLI API. I'll finish it over the weekend. Thanks!

niba committed:
feat(cli): add option --list to install command, remove list command

niba commented:
I refactored the code and implemented the logic as a ⁠`--list` option. I decided to remove some unnecessary new lines from the output because it can be a bit bloated if you have many instances on your computer. This is how it looks:

![image](https://github.com/user-attachments/assets/aa9f084b-6c1f-4851-9ab1-019e424cbed7)

pavelfeldman commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1009,6 +1024,50 @@ export class Registry {
       return await installDependenciesLinux(targets, dryRun);
   }
 
+  async list() {
```
I'm not sure this is what @dgozman means, but in systems with garbage collection (and we gc unused browsers), traversal and garbage collections are tightly coupled. We should have single traversal logic used in both. Listing installed browsers would be equivalent to running GC in a trial mode (i.e. without deleting the unused browsers). I.e. we already have all the code we need for the traversal, don't add anything new. As a bonus, you can print where each browser is installed.

niba replied:
Thanks, I got it. I don't remember exactly why I didn't use the traversal logic from ⁠`validateInstallationCache`, but I assume it was because in ⁠`validateInstallationCache` you are only interested in browser paths. In the `list` I try to group everything by project / installation and get additional information like browser version.

If I should reuse the traversal logic and don't change/add anything, then I'm not able to achieve the output that I showed.

I will try to unify the traversal, and you can decide if that's something you want. If not, then we can revert it.

btw I already print the location of every browser

niba committed:
feat(cli): unify traversing logic

niba committed:
feat(cli): unify traversing logic - part2

niba commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1237,40 +1281,50 @@ export class Registry {
+    return { linkTarget, descriptors: result };
+  }
+
   private async _validateInstallationCache(linksDir: string) {
```
I updated the code. ⁠`_traverse` returns all valid browsers for `⁠linkDir`, grouped by ⁠`linkTargets` (I need ⁠`linkTarget` in the list to print browsers per project).

I'm not quite sure about ⁠`_delete`. If we want to extract it, maybe it is better to call it `⁠_deleteAll`? The second section deletes all browsers except the ones you pass as a list (⁠`usedBrowsers` in this specific case).
So, the API could be ⁠`_deleteAll(browsersToPreserve: string[])`

pavelfeldman commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1237,12 +1265,13 @@ export class Registry {
-    for (const fileName of await fs.promises.readdir(linksDir)) {
-      const linkPath = path.join(linksDir, fileName);
+  private async _traverse(linksDir: string) {
+    const linkPaths = (await fs.promises.readdir(linksDir).catch(() => [])).map(link => path.join(linksDir, link));
```
`catch(() => [])` is a change to behavior here. Let's make sure the code stays exactly as it was. Including computing joined link inside the loop.

niba replied:
I took it from `uninstall` method. There, you use a catch while reading ⁠`linkDir`. I thought it could be useful here because you can call `⁠npx playwright install --list` when you haven't installed any browsers. I don't know if, in this case, `⁠linkDir` exists.

I will revert it

pavelfeldman commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1266,13 +1296,26 @@ export class Registry {
+            browserDescriptors.push(descriptor);
         }
+
+        result.push({ linkTarget, browsersDescriptors: browserDescriptors });
```
It does not make much sense to group them by the target - you end up listing the same browser over and over to the user. You should rather group them by the browser and list packages that are using the browser under each browser.

niba replied:
I know, but I have not received any negative comments about the output format, except for some verbosity.

I did it this way because the output format was not set, and it makes more sense to me to display browsers per project. I assumed this command is handy if you want to find out the exact browsers and versions your project uses, and this format lets you quickly grab that info.

I will change it to group by browsers.

pavelfeldman commented on packages/playwright-core/src/server/registry/index.ts:
```diff
@@ -1266,13 +1296,26 @@ export class Registry {
+    // 1. Collect used downloads and package descriptors.
+    const usedBrowserPaths: Set<string> = new Set();
+    const traverseData = await this._traverse(linksDir);
+    traverseData.forEach(({ browsersDescriptors }) => {
```
nit: sryle-wise we prefer for of loops.

pavelfeldman commented:
You are making good progress, but unfortunately the effort to review this work is greater than the effort to implement it. We assume high level or familiarity with the code and less iteration for the third-party contribution. I'll close it at this point. I understand it is frustrating, but please understand that we need to prioritize our time contributing to the project.
